Remove deprecated Predef.error

janekdb opened the pull request:
This was deprecated in 2.9.0.

I posted a question about the impact of this on `sbt` here: https://groups.google.com/forum/#!topic/sbt-dev/BNA692wX7Sk. Probably worth checking that thread in a few days - no objections so far but that could be because the post just hasn't be viewed widely enough yet.
## TODO
- [x] Wait for #5009 to be merged
- [x] Wait for #5049 to be merged

retronym commented:
The build failure is checking whether SBT 0.12.4's compiler interface can still be compiled. You can bump `sbt.latest.version` to something more recent in `build.xml` to reset this test to something more modern.

janekdb commented:
@retronym - Thanks. I've updated to the latest sbt release, temporarily dropped the removal of `error` and pushed to see the outcome of that change in isolation.

What command would I use to run the binary compatibility check locally?

adriaanm commented:
Looks like they moved to http://dl.bintray.com/typesafe/ivy-releases/org.scala-sbt/compiler-interface/0.13.11/srcs/compiler-interface-sources.jar

adriaanm commented:
We've been talking for a while now about creating a dedicated project in scala/scala that has sbt's compiler-specific logic, and coming up with a stable api. FYI, here's one of our discussions on that: https://github.com/scala/scala/compare/2.12.x...adriaanm:sbt-api-consolidate#commitcomment-16551474

SethTisue commented:
@janekdb let's wait to merge this until we can get all the way to 0.13.11

janekdb commented:
Sure.

janekdb committed:
Remove deprecated Predef.error

error was deprecated in 2.9.0 but remained to ensure compatibility with sbt.

This changes follows on from an update to the latest sbt version (0.13.11).

janekdb commented:
Review by @SethTisue, @adriaanm.

SethTisue commented:
LGTM

for the most part we are waiting until 2.13 to remove deprecated stuff, but this seems to me like a reasonable exception to make: it's been deprecated for donkey's years, only the sbt issue was holding up removal, it's only a forwarder, and getting things out of Predef is always especially attractive.

SethTisue commented:
/nothingtoseehere

SethTisue commented:
as I was merging this, @adriaanm raised a doubt about whether it's actually definitely OK to drop sbt 0.12 support in Scala 2.12 — it's conceivable those circles in the Venn diagram of the community might actually intersect? though one certainly hopes not...

but I thought about it more afterwards, and I don't think it matters, because sbt 0.12 uses Scala 2.9, which doesn't work on Java 8 anyway, and 2.12 is 8-only, so I think the circles _can't_ intersect

eed3si9n commented:
LGTM too. I don't know anyone who uses sbt 0.12 actively, and I don't think it's unreasonable to tell them to upgrade to sbt 0.13+ if they want to use Scala 2.12.

SethTisue commented:
@eed3si9n so then the other question is, do you happen to know exactly when sbt 0.13 became compatible with this change? both Adriaan and I were looking at commits in the sbt history that replaced `error` with `sys.error`, but it wasn't totally obvious to us which commits were relevant. is only the `compile/interface` directory relevant? code in that directory defines its own method named `error`, so it isn't immediately obvious where it might actually be calling `Predef.error`.

SethTisue commented:
@adriaanm I don't think this needs to be release-noted, even.

adriaanm commented:
Cool! Thanks for doubling down on the due diligence.

eed3si9n commented:
@SethTisue Yes https://github.com/sbt/sbt/tree/0.12/compile/interface should be the only relevant bit.
I was looking for it too, and found `def error(msg: String) = throw new RuntimeException(msg)` myself here - https://github.com/sbt/sbt/blob/0.12/compile/interface/API.scala#L23
